TechLadies/irrc-madwish
# IRRC-29. Implements Auth on backend (not enabled yet).

# Authentication


[IRRC-29](https://bootcamp-irrc.atlassian.net/browse/IRRC-29)


## Description

This change implements the backend needed for signup, login and protecting API routes against unauthenticated users.


## Test Steps

Before anything, add a value to your `.env` file.
```
...
JWT_KEY=randomString
```
You will also need to `npm install` and run the migrations.

1. Visit `/api/users/signup` with a POST request and a body of:
```json
{
  "Username": "my-username",
  "Password": "my-password"
}
```
This will sign you up.
2. Visit `/api/users/login` with a POST request and the same body as above. This will return you a `token`. You will use this token to access the other endpoints when we set them up to not be public.

Currently, this is all you can test. Later on, we will set up all the routes so that you cannot access them unless you pass a token. If you want to try this now first, then edit `/backend/src/app.js` as follows:

1. Choose a route that you would like to protect. Let's assume it's the `/students` route.
2. Make the following edit:
```js
app.use("/api/students", passport.authenticate('jwt', { session: false }), studentsRouter);
```
3. Now when you visit `/api/students`, you will get a Forbidden error.
4. Get a token from the `/user/login` endpoint.
5. Make the students call again, but attach the token:
<img width="968" alt="Screenshot 2021-03-24 at 6 19 40 PM" src="https://user-images.githubusercontent.com/21194045/112294334-d6829b00-8ccd-11eb-9a14-b2ac1ff57459.png">
6. It should return the students data this time.
